### DISCUSSION TITLE
libx11 fails to build due to wrong PKG_CONFIG path

### DISCUSSION THREAD

**Discussion by robske110** [Homebrew/discussions#4107] [Everyday usage] [answered]
### Output of `brew config`

```shell
HOMEBREW_VERSION: 3.6.18-42-g4abc51f
ORIGIN: https://github.com/Homebrew/brew
HEAD: 4abc51f32869511eaa29c7feaa30312aec70e1d2
Last commit: 2 hours ago
Core tap ORIGIN: https://github.com/Homebrew/homebrew-core
Core tap HEAD: 54cfefe1e8bf15282d162b02e90695cf7651382f
Core tap last commit: 27 minutes ago
Core tap branch: master
HOMEBREW_PREFIX: /usr/local
HOMEBREW_CASK_OPTS: []
HOMEBREW_MAKE_JOBS: 16
Homebrew Ruby: 2.6.8 => /usr/local/Homebrew/Library/Homebrew/vendor/portable-ruby/2.6.8_1/bin/ruby
CPU: 16-core 64-bit kabylake
Clang: 12.0.0 build 1200
Git: 2.24.3 => /Applications/Xcode.app/Contents/Developer/usr/bin/git
Curl: 7.64.1 => /usr/bin/curl
macOS: 10.15.7-x86_64
CLT: 12.0.0.32.29
Xcode: 12.4
```

### Output of `brew doctor`

```shell
Warning: You are using macOS 10.15.
We (and Apple) do not provide support for this old version.
It is expected behaviour that some formulae will fail to build in this old version.
It is expected behaviour that Homebrew will be buggy and slow.
Do not create any issues about this on Homebrew's GitHub repositories.
Do not create any issues even if you think this message is unrelated.
Any opened issues will be immediately closed without response.
Do not ask for help from MacHomebrew on Twitter.
You may ask for help in Homebrew's discussions but are unlikely to receive a response.
Try to figure out the problem yourself and submit a fix as a pull request.
We will review it but may or may not accept it.

Warning: Homebrew's "sbin" was not found in your PATH but you have installed
formulae that put executables in /usr/local/sbin.
Consider setting your PATH for example like so:
  echo 'export PATH="/usr/local/sbin:$PATH"' >> ~/.zshrc
```

### Description of issue

Installing libx11 while building from source fails with this:
```
configure: error: Package requirements (xproto >= 7.0.25 xextproto xtrans xcb >= 1.11.1 kbproto inputproto) were not met:

Package 'pthread-stubs', required by 'xcb', not found

Consider adjusting the PKG_CONFIG_PATH environment variable if you
installed software in a non-standard prefix.
```
I managed to **fix** this by adding ``ENV["PKG_CONFIG_PATH"] = `pkg-config --variable pc_path pkg-config` `` after this line: https://github.com/Homebrew/homebrew-core/blob/master/Formula/libx11.rb#L26
But I'm pretty sure that's just a bandaid fix and not the actual problem, which is why I have not made a PR with that change yet.
Does anyone know why pkg-config suddenly doesn't know the default paths anymore inside the configure script?

**Comment by gromgit** [comment, marked as answer]
As the error says, the problem's really with `libxcb`'s config script requiring `libpthreads-stubs`. However, [that's already addressed in libxcb.rb](https://github.com/Homebrew/homebrew-core/blob/54cfefe1e8bf15282d162b02e90695cf7651382f/Formula/libxcb.rb#L25-L27), and I'm currently rebuilding `libx11` on Catalina without the problems you're seeing.

What's the output of `brew info libxcb`?

**Reply by robske110** [reply]
The output of libxcb is:
```
==> libxcb: stable 1.15
X.Org: Interface to the X Window System protocol
https://www.x.org/
/usr/local/Cellar/libxcb/1.15 (2,459 files, 6.9MB) *
  Poured from bottle on 2022-11-01 at 12:20:22
From: https://github.com/Homebrew/homebrew-core/blob/HEAD/Formula/libxcb.rb
License: MIT
==> Dependencies
Build: pkg-config ✔, python@3.11 ✔, xcb-proto ✘
Required: libxau ✔, libxdmcp ✔
==> Analytics
install: 201,655 (30 days), 658,741 (90 days), 2,499,819 (365 days)
install-on-request: 619 (30 days), 1,748 (90 days), 8,987 (365 days)
build-error: 0 (30 days)
```

**Reply by gromgit** [reply]
Ah, that explains it. The patch commit I referenced earlier is dated Dec 27 2022, almost two months after you installed `libxcb`.

I'm afraid you'll have to:
```
brew rm --ignore-dependencies libxcb
brew install -s libxcb
```
then try to `brew install -s libx11` again.

**Reply by robske110** [reply]
A `brew reinstall libxcb` fixed the issues, and I can now build libx11 and related packages without my PKG_CONFIG modification. I must say I don't really understand why and how, but thank you very much for pointing out the real problem.
I had the same problem on a seperate catalina system (and same output of `brew info libxcb`), so I think it is unlikely that I accidentally removed xcb-proto.
If you can explain a bit more what you think the problem was (and why setting the PKG_CONFIG path "fixed it"), I'd be interested to read it.

**Reply by robske110** [reply]
Sorry, didn't see your comment, we posted at almost the same time.

**Reply by gromgit** [reply]
> If you can explain a bit more what you think the problem was (and why setting the PKG_CONFIG path "fixed it"), I'd be interested to read it.

In a nutshell:
1. Homebrew generates a custom `PKG_CONFIG_PATH` at build time, based _only_ on all the formula's _declared_ dependencies. This is to minimize the possibility of the build picking up unwanted libraries and stuff that just happen to be installed for other reasons ("opportunistic linkage" in Homebrew-speak).
2. Current release `xcb`'s (the upstream source, not the formula) config script quietly wants `pthread-stubs` (supplied by the `libpthread-stubs` formula) for both Linux and macOS.
3. _Development_ `xcb` [patches it out for macOS](https://gitlab.freedesktop.org/xorg/lib/libxcb/-/commit/d53ad60d774176c5b057e2ed4d03a4658ee13690), so the Homebrew maintainers [added the same patch](https://github.com/Homebrew/homebrew-core/pull/111928) to the `libxcb` formula, pending the next upstream release.

As for your setting `PKG_CONFIG_PATH` in the `libx11` formula, it only worked because **all** of the following were true:
1. `pkg-config --variable pc_path pkg-config` spits out the _base_ Homebrew pkgconfig directory paths, into which most installed formulae would symlink their `.pc` files. (The exceptions are those "keg-only" formulae which aren't linked into the base Homebrew hierarchy at all, usually because of conflicts with other formulae or base system software.)
2. You had `libpthread-stubs` already installed as a dependency of another formula, or for some other reason.
3. `libpthread-stubs` is _not_ keg-only, so its `.pc` files were symlinked into the base Homebrew pkgconfig directory.

Otherwise, the problem would keep recurring, until you rebuilt `libxcb`. You'd also be left with a `libxcb` installation that's _quietly_ dependent on `libpthread-stubs`, so Homebrew wouldn't stop you from removing the latter...and getting related errors later on.

**Comment by wyatt-wong** [comment]
I am using macOS Catalina 10.15.7 due to it is the latest macOS supported in my Retina MacBook Pro 2012. I encountered the SAME problem when I tried to upgrade `libx11.` After running `brew reinstall libxcb`, I was able to upgrade `libx11` to the latest version, thank you very much.

So the problem of failed to install or upgrade `libx11` was because the older version of `libxcb` was dependent on `libpthread-stubs`. But Homebrew maintainers have remove the dependency of `libpthread-stubs` from `libxcb` and so executing `brew reinstall libxcb` will reinstall the latest `libxcb` without the dependency of `libpthread-stubs` as well.

But why the dependency of `libpthread-stubs` is a problem for installing/upgrading `libx11` ?

**Reply by glensc** [reply]
I think libxcb needs revision bump (`revision 1`) for this to go away.
